Documentation/disaster-recovery.md

```diff
@@ -0,0 +1,112 @@
+  to this node.
+
+A Pod Checkpointer DaemonSet is deployed by default when using `bootkube
+render` to create cluster manifests. Using the Pod Checkpointer is highly
+recommended for all self-hosted clusters to ensure node reboot resiliency.
+
+For more information, see the [Pod Checkpointer
+README](https://github.com/kubernetes-incubator/bootkube/blob/master/cmd/checkpoint/README.md).
+
+## Bootkube Recover
+
+In the event of partial or total self-hosted control plane loss, `bootkube
+recover` may be able to assist in re-bootstrapping the self-hosted control
+plane.
+
+The `bootkube recover` subcommand does not recover a cluster directly. Instead,
+it extracts the control plane configuration from an available source and
+renders manifests in a format that `bootkube start` can use invoked to reboot
+the cluster.
+
+To see available options, run:
+
+```
+bootkube recover --help
+```
+
+To recover a cluster, first invoke `bootkube recover` with flags corresponding
+to the current state of the cluster (supported states listed below). Then,
+invoke `bootkube start` to reboot the cluster. For example:
+
+```
+scp bootkube user@master-node:
+ssh user@master-node
+./bootkube recover --asset-dir=recovered [scenario-specific options]
+sudo ./bootkube start --asset-dir=recovered
+```
+
+For complete recovery examples see the
+[hack/multi-node/bootkube-test-recovery](https://github.com/kubernetes-incubator/bootkube/blob/master/hack/multi-node/bootkube-test-recovery)
+and
+[hack/multi-node/bootkube-test-recovery-self-hosted-etcd](https://github.com/kubernetes-incubator/bootkube/blob/master/hack/multi-node/bootkube-test-recovery-self-hosted-etcd)
+scripts. The `bootkube-test-recovery` script is demoed below.
+
+[![asciicast](https://asciinema.org/a/dsp43ziuuzwcztni94y8l25s5.png)](https://asciinema.org/a/dsp43ziuuzwcztni94y8l25s5)
+
+### If an api-server is still running
+
+If an api-server is still running but other control plane components are down,
+preventing cluster functionality (i.e. the scheduler pods are all down), the
+control plane can be extracted directly from the api-server:
+
+```
+bootkube recover --asset-dir=recovered --kubeconfig=/etc/kubernetes/kubeconfig
+```
+### If an external etcd cluster is still running
+
+If using an external (non-self-hosted) etcd cluster, the control plane can be
+extracted directly from etcd:
+
+```
+bootkube recover --asset-dir=recovered --etcd-servers=http://127.0.0.1:2379 --kubeconfig=/etc/kubernetes/kubeconfig
+```
+
+### If an etcd backup is available (non-self-hosted etcd)
+
+First, recover the external etcd cluster from the backup. Then use the method
+described in the previous section to recover the control plane manifests.
+
+### If an etcd backup is available (self-hosted etcd)
+
+If using self-hosted etcd, recovery is supported via reading from an etcd
+backup file:
+
+```
+bootkube recover --asset-dir=recovered --etcd-backup-file=backup --kubeconfig=/etc/kubernetes/kubeconfig
+```
+
```

Comment:
we tried this with our field engs yesterday. 

there are a few things we need to make sure before run this script:

1. kubelet is running on the machine
2. no related containers are running (old etcd, old api server, etc.. this is also applied to other recovery cases i believe)
3. docker state is clean (docker ps -a does not contain old states of relevant containers). kubelet has bugs that it might incorrectly believes the static pod has dead when old state exists.
4. /var/etcd dir is clean on ALL master nodes